File: typescript/iterator/book-shelf-iterator.ts
import { Iterator } from "./iterator";
import { BookShelf } from "./book-shelf";
import { Book } from "./book";

export class BookShelfIterator implements Iterator<Book> {
  private index = 0;

  constructor(private bookShelf: BookShelf) {}

  hasNext(): boolean {
    if (this.index < this.bookShelf.length) {
      return true;
    }
    return false;
  }

  next(): Book {
    const book = this.bookShelf.getBookAt(this.index);
    this.index++;
    return book;
  }
}

File: kotlin/iterator/src/BookShelfIterator.kt
package iterator

class BookShelfIterator(private val bookShelf: BookShelf) : Iterator<Book> {
    private var index = 0

    override fun hasNext(): Boolean {
        if (index < bookShelf.length) {
            return true
        }
        return false
    }

    override fun next(): Book {
        val book = bookShelf.getBookAt(index)
        this.index++
        return book as Book
    }
}